src/nnet3bin/nnet3-copy-multiple-egs.cc

```diff
@@ -0,0 +1,496 @@
+                   NnetExample *eg_out) {
+  eg_out->io.clear();
+  eg_out->io.resize(eg.io.size());
+  for (size_t i = 0; i < eg.io.size(); i++) {
+    bool is_input_or_output;
+    int32 min_t, max_t;
+    const NnetIo &io_in = eg.io[i];
+    NnetIo &io_out = eg_out->io[i];
+    const std::string &name = io_in.name;
+    io_out.name = name;
+    if (name == "input") {
+      min_t = min_input_t;
+      max_t = max_input_t;
+      is_input_or_output = true;
+    } else if (name == "output") {
+      min_t = min_output_t;
+      max_t = max_output_t;
+      is_input_or_output = true;
+    } else {
+      is_input_or_output = false;
+    }
+    if (!is_input_or_output) {  // Just copy everything.
+      io_out.indexes = io_in.indexes;
+      io_out.features = io_in.features;
+    } else {
+      const std::vector<Index> &indexes_in = io_in.indexes;
+      std::vector<Index> &indexes_out = io_out.indexes;
+      indexes_out.reserve(indexes_in.size());
+      int32 num_indexes = indexes_in.size(), num_kept = 0;
+      KALDI_ASSERT(io_in.features.NumRows() == num_indexes);
+      std::vector<bool> keep(num_indexes, false);
+      std::vector<Index>::const_iterator iter_in = indexes_in.begin(),
+                                          end_in = indexes_in.end();
+      std::vector<bool>::iterator iter_out = keep.begin();
+      for (; iter_in != end_in; ++iter_in,++iter_out) {
+        int32 t = iter_in->t;
+        bool is_within_range = (t >= min_t && t <= max_t);
+        *iter_out = is_within_range;
+        if (is_within_range) {
+          indexes_out.push_back(*iter_in);
+          num_kept++;
+        }
+      }
+      KALDI_ASSERT(iter_out == keep.end());
+      if (num_kept == 0)
+        KALDI_ERR << "FilterExample removed all indexes for '" << name << "'";
+
+      FilterGeneralMatrixRows(io_in.features, keep,
+                              &io_out.features);
+      KALDI_ASSERT(io_out.features.NumRows() == num_kept &&
+                   indexes_out.size() == static_cast<size_t>(num_kept));
+    }
+  }
+}
+
+
+/**
+   This function is responsible for possibly selecting one frame from multiple
+   supervised frames, and reducing the left and right context as specified.  If
+   frame == "" it does not reduce the supervised frames; if frame == "random" it
+   selects one random frame; otherwise it expects frame to be an integer, and
+   will select only the output with that frame index (or return false if there was
+   no such output).
+
+   If left_context != -1 it removes any inputs with t < (smallest output - left_context).
+      If left_context != -1 it removes any inputs with t < (smallest output - left_context).
+
+   It returns true if it was able to select a frame.  We only anticipate it ever
+   returning false in situations where frame is an integer, and the eg came from
+   the end of a file and has a smaller than normal number of supervised frames.
+
+*/
+bool SelectFromExample(const NnetExample &eg,
```

Comment:
If this function is the same as in nnet3-copy-egs, you should move both of these files to nnet3/nnet-example-utils.cc.